[TF-TRT] Fix native fallback issue in StridedSlice with dynamic input indices and shrink_axis_mask

pavanimajety opened the pull request:
This PR introduces two changes: 

1. Returns unimplemented error when begin tensor has negative values.
2. Fixes the logic when shrink_axis mask is used with unknown input dimensions.

pavanimajety commented:
@tfeher @DEKHTIARJonathan Please review

tfeher commented on tensorflow/compiler/tf2tensorrt/convert/ops/slice_ops.cc:
```diff
@@ -152,6 +176,36 @@ Status HandleDynamicStridedSliceInput(
 
+  // When begin_mask or end_mask are set, we have to disregard the begin_tensor
+  // and end_tensor values. In static indices cases, ValidateStridedSliceOp
+  // returns the correct begin_tensor and end_tensor values, howe with dynamic
```
```suggestion
  // returns the correct begin_tensor and end_tensor values, however with dynamic
```

tfeher commented on tensorflow/compiler/tf2tensorrt/convert/ops/slice_ops.cc:
```diff
@@ -130,11 +134,31 @@ Status ConvertStridedSliceHelper(
+                      params, &tensor, op_instance));
+   }
+   else {
+        TF_RETURN_IF_ERROR(PrepareTensorForShape(
```
Is this else branch necessary? I believe `new_axis_mask `is not supported, therefore we only need reshape if we are squeezing away axis. (Is that correct?) And `PrepareTensorForShape `would not work for dynamic input dims.

pavanimajety replied:
Currently tests related to Unpack and Split fail when I remove the if-else. Is it okay if I take this as a feedback for my upcoming PR?

tfeher replied:
Yes, it  is fine to address this in the next PR.

tfeher commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -4961,6 +4961,159 @@ TEST_P(OpConverter_FP32_Test, ConvertSqueeze) {
+         int64 ellipsis_mask = 0, int64 new_axis_mask = 0,
+         int64 shrink_axis_mask = 0) -> NodeDef {
+    Scope s = Scope::NewRootScope();
+    auto input = ops::Placeholder(s.WithOpName("input"), tf_type);
+    auto begin = ops::Placeholder(s.WithOpName("begin"), DT_INT32);
+    auto end = ops::Placeholder(s.WithOpName("end"), DT_INT32);
+    auto strides = ops::Placeholder(s.WithOpName("strides"), DT_INT32);
+    ops::StridedSlice::Attrs attrs = ops::StridedSlice::Attrs()
+                                         .BeginMask(begin_mask)
+                                         .EndMask(end_mask)
+                                         .EllipsisMask(ellipsis_mask)
+                                         .NewAxisMask(new_axis_mask)
+                                         .ShrinkAxisMask(shrink_axis_mask);
+    auto strided_slice = ops::StridedSlice(s.WithOpName("my_strided_slice"),
+                                           input, begin, end, strides, attrs);
+    return strided_slice.operation.node()->def();
+  };
+  struct TestParams {
+    std::vector<int> input_dims;
+    std::vector<int> begin;
+    std::vector<int> end;
+    std::vector<int> strides;
+    int begin_mask;
+    int end_mask;
+    int ellipsis_mask;
+    int new_axis_mask;
+    int shrink_axis_mask;
+    std::vector<int> expected_output_dims;
+    std::vector<float> expected_output;
+    Status conversion_status;
+    Status runtime_status;
+    std::vector<int> partial_input_dims;
+  };
+  auto get_mask = [](const std::vector<int>& mask) {
+   int result = 0;
+   for (int i = 0; i < mask.size(); i++) {
+     if (mask[i]) result += (1 << i);
+   }
+   return result;
+ };
```
Please remove code duplication: you should reuse code for ConvertStridedSlice and ConvertStridedSlice2 tests.

tfeher commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -4961,6 +4961,159 @@ TEST_P(OpConverter_FP32_Test, ConvertSqueeze) {
+   VLOG(2) << "Preparing test case " << i++ << " with dims "
+           << DebugString(p.input_dims);
+
+   switch (trt_mode_) {
+     case TrtTestMode::kImplicitBatch: {
+       AddTestTensor("input", p.input_dims, ok_input);
+       break;
+     }
+     case TrtTestMode::kExplicitBatch: {
+       AddTestTensor("input", p.input_dims, ok_input);
+       break;
+     }
+     case TrtTestMode::kDynamicShape: {
+       if (p.partial_input_dims.size() > 0) {
+         AddTestTensor("input", p.input_dims, tf_type_, ok_input,
+                       p.partial_input_dims);
+
+       } else {
+         AddTestTensor("input", p.input_dims, tf_type_, ok_input,
+                       p.input_dims);
+       }
+       break;
+     }
+   }
```
Could be more concise
```suggestion
       if (if trt_mode == TrtTestMode::kDynamicShape &&  p.partial_input_dims.size() > 0) {
         AddTestTensor("input", p.input_dims, tf_type_, ok_input,
                       p.partial_input_dims);
       } else {
         AddTestTensor("input", p.input_dims, ok_input); 
       }
```

tfeher reviewed (changes requested):
Thanks @pavanimajety for the PR! It looks good overall, I have just a few comments.

tfeher reviewed (approved):
Thanks Pavani for the updates, it looks good to me!

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/slice_ops.cc:
```diff
@@ -152,6 +176,36 @@ Status HandleDynamicStridedSliceInput(
+  nvinfer1::Dims dims = input_tensor->getDimensions();
+
+  for (int idx = 0; idx < dims.nbDims; ++idx){
+    // VLOG(2) << "begin_mask[" << idx << "]: "<< begin_mask[idx];
+    // VLOG(2) << "end_mask[" << idx << "]: "<< end_mask[idx];
+    // VLOG(2) << "shrink_mask[" << idx << "]: "<< shrink_axis_mask[idx];
```
What do you want to do with these lines?

pavanimajety replied:
These are optional debugging statements. I wanted to keep them commented because slice_ops.cc seems to be a place where we debug for many operations like split, unpack, slice, strided slice etc. and these lines with mask properties just make it simpler to debug. I can remove them if they are better removed.

bixia1 replied:
It is either in or out, we can keep debug code for this. You may want to make them VLOG(3).

pavanimajety replied:
Done

pavanimajety committed:
Fix native fallback issue in StridedSlice

Return unimplemented error when begin tensor has dynamic values and fix the logic when shrink_axis mask is used with unknown input dimensions.
Fix native fallback with strided slice op.


Fix compilation errors


Fix compiler errors: 2


Fix unit tests


Fix issue when end_mask and shrink_axis_mask are both enabled.


Address Tamas' feedback -- Combining tests, minor changes.


Remove unused variable


Modify the logging levels for logs


Fix indentation
